cloud/aws/aws_env.cc

```diff
@@ -221,38 +222,86 @@ Aws::S3::Model::HeadObjectOutcome AwsS3ClientWrapper::HeadObject(
   return outcome;
 }
 
+static bool GetNameFromEnvironment(const char *name, const char *alt, std::string * result) {
+  char *value = getenv(name);
+  if (value == nullptr && alt != nullptr) {
+    value = getenv(alt);
+  }
+  if (value != nullptr) {
+    result->assign(value);
+    return true;
+  } else {
+    return false;
+  }
+}
+  
+static std::string GetFullBucketName(const std::string & name, const char *bucket) {
+  std::string full_name;
+  if (! GetNameFromEnvironment(bucket, "ROCKSDB_CLOUD_BUCKET_NAME", &full_name)) {
```

Comment:
I would prefer not use environment variable to control a thing such as bucket name.

What do you think about the following: add an option in CloudEnvOptions `bucket_name_add_rockset_prefix`, defaults to true for backwards compatibility. You can then set it to `false` in your environment.

Reply:
We *really* want to control all of RocksSet without using hard-coded parameters.  We hope to use RocksDB as a bundle plugged into software we do not compile or own.  That is the point of most of our changes.

In the future we hope to be able to control things like bucket name via INI files (as well as possibly the environment).  But the environment is our only option at the moment.

What is the object to controlling some of the parameters through the environment?

Reply:
@mrambacher I agree with you that you want to fully control the bucket path. What I'm saying is in line with that -- a parameter given to `CloudEnv::NewAwsEnv()` needs to be full bucket name and RocksDB-cloud should not add *any* prefix to it. Right now it is adding a prefix that is controlled by the environment variable, but I think it should not even do that. Don't add any prefix, period.

Do you agree with this?

However, we'll unfortunately need to keep the backwards compatibility, which adds `rockset.` as a prefix to the bucket name. Let's control this behavior using `CloudEnvOptions`. Does this makes sense?

> But the environment is our only option at the moment.

It is fine if your code controls the bucket name using the environment variable. But you should read the environment variable in the binary and provide the bucket_name value in `CloudEnv::NewAwsEnv()`.

Reply:
The problem is that I do not necessarily have control of the binary going forward.   I want to integrate the Cloud code into Rocksandra or MyRocks.  I do not want to modify the code of those products to set additional parameters to specify bucket names or prefixes.  Ideally, I want to get rid of the options to NewAwsEnv, not add more.

I also believe there are probably other examples of code trying to replace the "rockset." prefix already.  In db_test_util, "dbtest." is prepended to the name of the bucket.  I would assume that the code would think buckets are named by this parameter and not rockset.dbtest."

If you really want, I can add a "prefix" to the CloudEnvOptions.  I can also add the other bucket parameters to CloudEnvOptions at the same time.  However, I still need a means of setting (or at least initializing) them to some other value via the environment.

Reply:
I think the rule of thumb should be that all configuration happens through `CloudEnvOptions`. We can initialize `CloudEnvOptions` through either environment variables or .ini files or whatever else. But once initialized, we should pass `CloudEnvOptions` to `NewAwsEnv()` and all behavior of RocksDB-Cloud should be evident by inspecting contents of `CloudEnvOptions`.

Does this make sense?

> If you really want, I can add a "prefix" to the CloudEnvOptions. I can also add the other bucket parameters to CloudEnvOptions at the same time.

Yes, let's do that.

> However, I still need a means of setting (or at least initializing) them to some other value via the environment.

Feel free to introduce a new function `InitializeCloudEnvFromEnvironment` or even `IntroduceCloudEnvFromConfigFile`. I do not know how we'll invoke this function from MyRocks, but whatever we do, we should have a clean separation. Would that work for you?

Reply:
> I also believe there are probably other examples of code trying to replace the "rockset." prefix already. In db_test_util, "dbtest." is prepended to the name of the bucket. I would assume that the code would think buckets are named by this parameter and not rockset.dbtest."

Yes, I do not like this transparent adding of `rockset.` prefix. If we migrate away from it, we will break backwards compatibility, though :(

Reply:
I pushed out another changeset to this PR today.  Because of its where we are in the review, I did not merge it into the other change set.  In this changeset, I added a BucketOptions class that contains the settings (bucket name, object prefix, region) for a bucket.  There are now two of these (source, dest) in a CloudEnvOptions.  When created, the bucket settings are initialized from the environment.  You can also explicitly override them via Set methods.  There is also an "Initialize" method that sets the values to the input ones if the input ones are not blank.

I also added another NewAwsEnv signature that has all of these values in the options. 

Please take a look at the update.  I have tested db_cloud_test and everything passes here.

Reply:
Overall I like the change. Two general comments:
1) `CloudEnv::NewAwsEnv` now takes bucket config from both the function arguments and also from `CloudEnvOptions`. Specifying one thing in two different ways is not a good API design. We should remove the function arguments. This will break the backwards compatibility, so we need to make sure to specify the change in release notes (HISTORY.md).
2) I am still not on board with the default constructor of `BucketOptions` reading environment variables. I would introduce `BucketOptions initializeSourceBucketFromEnviroment()` and `BucketOptions intializeDestinationBucketFromEnvironment()`. Given the today's design of RocksDB-Cloud, we will need to have a different initialization path that we include in MyRocks and Rocksandra. We can call these two functions from there (in the future).